From 435286250304a29806d3e1abc0926dd65f4a276b Mon Sep 17 00:00:00 2001
From: Jiuyang Liu <liu@jiuyang.me>
Date: Sun, 24 Apr 2022 23:46:06 +0800
Subject: [PATCH] chisel plugin should also be applied to chisel3

---
 build.sc | 12 +++++-------
 1 file changed, 5 insertions(+), 7 deletions(-)

diff --git a/build.sc b/build.sc
index 09f5ceebf9..64bd4d2cf5 100644
--- a/build.sc
+++ b/build.sc
@@ -104,12 +104,12 @@ class chisel3CrossModule(val crossScalaVersion: String) extends CommonModule wit
     )
   }
 
-  object test extends Tests with TestModule.ScalaTest {
-    override def scalacPluginClasspath = m.scalacPluginClasspath
+  override def scalacOptions = T {
+    super.scalacOptions() ++ Agg(s"-Xplugin:${plugin.jar().path}", "-P:chiselplugin:genBundleElements")
+  }
 
-    override  def scalacOptions = T {
-      super.scalacOptions() ++ Agg("-P:chiselplugin:genBundleElements")
-    }
+  object test extends Tests with TestModule.ScalaTest {
+    override def scalacPluginClasspath = T { m.scalacPluginClasspath() }
 
     override def ivyDeps = m.ivyDeps() ++ Agg(
       v.scalatest,
@@ -120,8 +120,6 @@ class chisel3CrossModule(val crossScalaVersion: String) extends CommonModule wit
   }
 
   object `integration-tests` extends Tests with TestModule.ScalaTest {
-    override def scalacPluginClasspath = m.scalacPluginClasspath
-
     override def sources = T.sources(millSourcePath / "integration-tests" / "src" / "test" / "scala")
     override def ivyDeps = m.ivyDeps() ++ Agg(
       v.scalatest,
