From e841e76f8c894f3b61aa6cacf54a04721d4bff2e Mon Sep 17 00:00:00 2001
From: Jiuyang Liu <liu@jiuyang.me>
Date: Sat, 22 May 2021 03:16:04 +0000
Subject: [PATCH] mill_fix

---
 build.sc | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/build.sc b/build.sc
index 8c385966..6b8a5870 100644
--- a/build.sc
+++ b/build.sc
@@ -54,7 +54,9 @@ class chiseltestCrossModule(val crossScalaVersion: String) extends CrossSbtModul
   override def ivyDeps = T {
     Agg(
       ivy"org.scalatest::scalatest:3.2.0",
-      ivy"com.lihaoyi::utest:0.7.9"
+      ivy"com.lihaoyi::utest:0.7.9",
+      ivy"org.json4s::json4s-jackson:3.6.11",
+      ivy"javax.xml.bind:jaxb-api:2.3.1",
     ) ++ chisel3IvyDeps ++ treadleIvyDeps
   }
 
@@ -62,7 +64,7 @@ class chiseltestCrossModule(val crossScalaVersion: String) extends CrossSbtModul
     override def ivyDeps = T {
       Agg(
         ivy"org.scalatest::scalatest:3.0.8",
-        ivy"com.lihaoyi::utest:0.7.9"
+        ivy"com.lihaoyi::utest:0.7.9",
       ) ++ chisel3IvyDeps ++ treadleIvyDeps
     }
 
