From 1c9fb855c7e0eca943c32bb7c6f1ba7c20659a20 Mon Sep 17 00:00:00 2001
From: Jiuyang Liu <liu@jiuyang.me>
Date: Tue, 17 Aug 2021 13:30:59 +0800
Subject: [PATCH] downgrade scalatest

---
 build.sc | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/build.sc b/build.sc
index a54798de..fe50822c 100644
--- a/build.sc
+++ b/build.sc
@@ -67,7 +67,7 @@ class chiseltestCrossModule(val crossScalaVersion: String) extends CrossSbtModul
 
   override def ivyDeps = T {
     Agg(
-      ivy"org.scalatest::scalatest:3.2.0",
+      ivy"org.scalatest::scalatest:3.1.4",
       ivy"com.lihaoyi::utest:0.7.9",
       ivy"net.java.dev.jna:jna:5.8.0",
     ) ++ chisel3IvyDeps ++ treadleIvyDeps ++ malteseIvyDeps
