From 024325f1dab234a137239097c85ad5a86aeb0302 Mon Sep 17 00:00:00 2001
From: Zenithal <i@zenithal.me>
Date: Sun, 6 Feb 2022 17:42:54 +0800
Subject: [PATCH] Eliminate Wunknown-pragmas for Clang/LLVM

---
 util/string.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/util/string.c b/util/string.c
index 90d7127e..9ce53fa2 100644
--- a/util/string.c
+++ b/util/string.c
@@ -4,7 +4,7 @@
 #include <stdint.h>
 #include <ctype.h>
 
-#ifdef __GNUC__
+#if defined(__GNUC__) && !defined(__llvm__)
 // Don't let GCC pattern-match these functions' bodies into self-calls
 #pragma GCC optimize ("no-tree-loop-distribute-patterns")
 #endif
