diff --git a/src/main/scala/DivSqrtRecFN_small.scala b/src/main/scala/DivSqrtRecFN_small.scala
index 32b5572..a900fc0 100644
--- a/src/main/scala/DivSqrtRecFN_small.scala
+++ b/src/main/scala/DivSqrtRecFN_small.scala
@@ -382,7 +382,7 @@ class
         processTwoBits && !newBit && (0.S < (rem_Z<<2)(sigWidth+2, 0).zext - trialTerm2_newBit0.zext) ||
         !(processTwoBits && newBit2) && nextNotZeroRem_Z
     val nextRem_Z_2 =
-        Mux(processTwoBits && newBit2,  trialRem2.asUInt()(sigWidth + 1, 0), 0.U) |
+        Mux(processTwoBits && newBit2,  trialRem2.asUInt(sigWidth + 1, 0), 0.U) |
         Mux(processTwoBits && !newBit2, rem2(sigWidth + 1, 0),               0.U) |
         Mux(!processTwoBits,            nextRem_Z,                           0.U)
 
