diff --git a/src/main/scala/gemmini/AccumulatorScale.scala b/src/main/scala/gemmini/AccumulatorScale.scala
index bf98a144..e4aaa192 100644
--- a/src/main/scala/gemmini/AccumulatorScale.scala
+++ b/src/main/scala/gemmini/AccumulatorScale.scala
@@ -160,7 +160,7 @@ class AccumulatorScale[T <: Data, U <: Data](
           regs(i).valid := false.B
         }
       }
-      head_oh := (head_oh << 1).asUInt() | head_oh(nEntries-1)
+      head_oh := (head_oh << 1).asUInt | head_oh(nEntries-1)
     }
 
     io.in.ready := !Mux1H(tail_oh.asBools, regs.map(_.valid)) || (tail_oh === head_oh && out.fire)
