From 861a6d824310647f2acd0c35e9143a3a949e6edd Mon Sep 17 00:00:00 2001
From: Jiuyang Liu <liu@jiuyang.me>
Date: Sat, 29 May 2021 00:29:55 +0000
Subject: [PATCH] remove testOnly, since upstream implemented it.

---
 common.sc | 6 ------
 1 file changed, 6 deletions(-)

diff --git a/common.sc b/common.sc
index 4b57052f8f5..370b42c5a15 100644
--- a/common.sc
+++ b/common.sc
@@ -57,12 +57,6 @@ trait CommonRocketChip extends SbtModule with PublishModule {
     def testFrameworks = T {
       Seq("org.scalatest.tools.Framework")
     }
-
-    // a sbt-like testOnly command.
-    // for example, mill -i "chisel3[2.12.12].test.testOnly" "chiselTests.BitwiseOpsSpec"
-    def testOnly(args: String*) = T.command {
-      super.runMain("org.scalatest.run", args: _*)
-    }
   }
 
   override def millSourcePath = super.millSourcePath / os.up
