diff --git a/build.sbt b/build.sbt
index 7749871a2be..ed7b40300af 100644
--- a/build.sbt
+++ b/build.sbt
@@ -15,7 +15,7 @@ lazy val commonSettings = Seq(
   traceLevel   := 15,
   scalacOptions ++= Seq("-deprecation","-unchecked"),
   libraryDependencies ++= Seq("org.scala-lang" % "scala-reflect" % scalaVersion.value),
-  libraryDependencies ++= Seq("org.json4s" %% "json4s-jackson" % "3.6.6"),
+  libraryDependencies ++= Seq("org.json4s" %% "json4s-jackson" % "4.0.5"),
   libraryDependencies ++= Seq("org.scalatest" %% "scalatest" % "3.2.0" % "test"),
   resolvers ++= Seq(
     Resolver.sonatypeRepo("snapshots"),
diff --git a/common.sc b/common.sc
index ed7b967184e..72ee299a4e8 100644
--- a/common.sc
+++ b/common.sc
@@ -84,7 +84,7 @@ trait CommonRocketChip extends SbtModule with PublishModule {
   override def ivyDeps = T {
     Agg(
       ivy"${scalaOrganization()}:scala-reflect:${scalaVersion()}",
-      ivy"org.json4s::json4s-jackson:3.6.6",
+      ivy"org.json4s::json4s-jackson:4.0.5",
       ivy"org.scalatest::scalatest:3.2.0"
     ) ++ chisel3IvyDeps
   }
