From c2195ce759f03a28577f0c4a3b9202efa4011191 Mon Sep 17 00:00:00 2001
From: Jiuyang Liu <liu@jiuyang.me>
Date: Sat, 8 Jan 2022 03:12:37 +0800
Subject: [PATCH] bump for bump for chipsalliance/rocket-chip#2841

---
 src/main/scala/CVA6Tile.scala | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/src/main/scala/CVA6Tile.scala b/src/main/scala/CVA6Tile.scala
index edca98f..b019041 100644
--- a/src/main/scala/CVA6Tile.scala
+++ b/src/main/scala/CVA6Tile.scala
@@ -41,6 +41,7 @@ case class CVA6CoreParams(
 ) extends CoreParams {
   /* DO NOT CHANGE BELOW THIS */
   val useVM: Boolean = true
+  val useHypervisor: Boolean = false
   val useUser: Boolean = true
   val useSupervisor: Boolean = false
   val useDebug: Boolean = true
@@ -74,6 +75,7 @@ case class CVA6CoreParams(
   val decodeWidth: Int = 1 // TODO: Check
   val fetchWidth: Int = 1 // TODO: Check
   val retireWidth: Int = 2
+  val nPTECacheEntries: Int = 8 // TODO: Check
 }
 
 case class CVA6TileAttachParams(
