From 13991a938d12c6430223ca6fe680029cf146c230 Mon Sep 17 00:00:00 2001
From: Jiuyang Liu <liu@jiuyang.me>
Date: Sat, 14 May 2022 02:07:26 +0800
Subject: [PATCH] fix for chipsalliance/rocket-chip#2925

---
 .../polarfire_pcie_rootport/PolarFirePCIeRootPort.scala       | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/main/scala/ip/microsemi/polarfire_pcie_rootport/PolarFirePCIeRootPort.scala b/src/main/scala/ip/microsemi/polarfire_pcie_rootport/PolarFirePCIeRootPort.scala
index 5657573..8f2266c 100644
--- a/src/main/scala/ip/microsemi/polarfire_pcie_rootport/PolarFirePCIeRootPort.scala
+++ b/src/main/scala/ip/microsemi/polarfire_pcie_rootport/PolarFirePCIeRootPort.scala
@@ -330,7 +330,7 @@ class PolarFirePCIeX4(implicit p:Parameters) extends LazyModule
 
     //m
     m.aw.bits.cache := UInt(0)
-    m.aw.bits.prot  := AXI4Parameters.PROT_PRIVILEDGED
+    m.aw.bits.prot  := AXI4Parameters.PROT_PRIVILEGED
     m.aw.bits.qos   := UInt(0)
 
     m.aw.bits.id                        := blackbox.io.PCIESS_AXI_1_M_AWID
@@ -354,7 +354,7 @@ class PolarFirePCIeX4(implicit p:Parameters) extends LazyModule
     m.b.ready                           := blackbox.io.PCIESS_AXI_1_M_BREADY
 
     m.ar.bits.cache := UInt(0)
-    m.ar.bits.prot  := AXI4Parameters.PROT_PRIVILEDGED
+    m.ar.bits.prot  := AXI4Parameters.PROT_PRIVILEGED
     m.ar.bits.qos   := UInt(0)
 
     m.ar.bits.id                        := blackbox.io.PCIESS_AXI_1_M_ARID
