diff --git a/src/main/scala/ip/microsemi/corejtagdebug/corejtagdebug.scala b/src/main/scala/ip/microsemi/corejtagdebug/corejtagdebug.scala
index 0cb2327..039c8d7 100644
--- a/src/main/scala/ip/microsemi/corejtagdebug/corejtagdebug.scala
+++ b/src/main/scala/ip/microsemi/corejtagdebug/corejtagdebug.scala
@@ -1,6 +1,7 @@
 package sifive.fpgashells.ip.microsemi.corejtagdebug
 
 import Chisel._
+import chisel3.BlackBox
 import chisel3.experimental.{Analog,attach}
 import freechips.rocketchip.util.{ElaborationArtefacts}
 import freechips.rocketchip.util.GenericParameterizedBundle
diff --git a/src/main/scala/ip/microsemi/polarfire.scala b/src/main/scala/ip/microsemi/polarfire.scala
index a304018..83bccc5 100644
--- a/src/main/scala/ip/microsemi/polarfire.scala
+++ b/src/main/scala/ip/microsemi/polarfire.scala
@@ -1,6 +1,7 @@
 package sifive.fpgashells.ip.microsemi
 
 import Chisel._
+import chisel3.BlackBox
 import chisel3.{Input, Output}
 import chisel3.experimental.{Analog, attach}
 import freechips.rocketchip.util.{ElaborationArtefacts}
diff --git a/src/main/scala/ip/microsemi/polarfire_ccc/PolarFireCCC.scala b/src/main/scala/ip/microsemi/polarfire_ccc/PolarFireCCC.scala
index bc740a0..26a411e 100644
--- a/src/main/scala/ip/microsemi/polarfire_ccc/PolarFireCCC.scala
+++ b/src/main/scala/ip/microsemi/polarfire_ccc/PolarFireCCC.scala
@@ -1,6 +1,7 @@
 package sifive.fpgashells.ip.microsemi.polarfireccc
 
 import Chisel._
+import chisel3.BlackBox
 import freechips.rocketchip.util.{ElaborationArtefacts}
 import org.chipsalliance.cde.config._
 
diff --git a/src/main/scala/ip/microsemi/polarfire_clock_divider/PolarFireClockDivider.scala b/src/main/scala/ip/microsemi/polarfire_clock_divider/PolarFireClockDivider.scala
index e515216..e665d87 100644
--- a/src/main/scala/ip/microsemi/polarfire_clock_divider/PolarFireClockDivider.scala
+++ b/src/main/scala/ip/microsemi/polarfire_clock_divider/PolarFireClockDivider.scala
@@ -1,6 +1,7 @@
 package sifive.fpgashells.ip.microsemi.polarfireclockdivider
 
 import Chisel._
+import chisel3.BlackBox
 import chisel3.experimental.{Analog,attach}
 import freechips.rocketchip.util.{ElaborationArtefacts}
 import freechips.rocketchip.util.GenericParameterizedBundle
diff --git a/src/main/scala/ip/microsemi/polarfire_ddr3/PolarFireDDR3.scala b/src/main/scala/ip/microsemi/polarfire_ddr3/PolarFireDDR3.scala
index b6e94b4..1b6235b 100644
--- a/src/main/scala/ip/microsemi/polarfire_ddr3/PolarFireDDR3.scala
+++ b/src/main/scala/ip/microsemi/polarfire_ddr3/PolarFireDDR3.scala
@@ -1,6 +1,7 @@
 package sifive.fpgashells.ip.microsemi.polarfireddr3
 
 import Chisel._
+import chisel3.BlackBox
 import chisel3.experimental.{Analog,attach}
 import freechips.rocketchip.util.{ElaborationArtefacts}
 import freechips.rocketchip.util.GenericParameterizedBundle
diff --git a/src/main/scala/ip/microsemi/polarfire_ddr4/PolarFireDDR4.scala b/src/main/scala/ip/microsemi/polarfire_ddr4/PolarFireDDR4.scala
index 7643a73..5cc2217 100644
--- a/src/main/scala/ip/microsemi/polarfire_ddr4/PolarFireDDR4.scala
+++ b/src/main/scala/ip/microsemi/polarfire_ddr4/PolarFireDDR4.scala
@@ -1,6 +1,7 @@
 package sifive.fpgashells.ip.microsemi.polarfireddr4
 
 import Chisel._
+import chisel3.BlackBox
 import chisel3.experimental.{Analog,attach}
 import freechips.rocketchip.util.{ElaborationArtefacts}
 import freechips.rocketchip.util.GenericParameterizedBundle
diff --git a/src/main/scala/ip/microsemi/polarfire_dll/PolarFireDLL.scala b/src/main/scala/ip/microsemi/polarfire_dll/PolarFireDLL.scala
index 6135590..01731c5 100644
--- a/src/main/scala/ip/microsemi/polarfire_dll/PolarFireDLL.scala
+++ b/src/main/scala/ip/microsemi/polarfire_dll/PolarFireDLL.scala
@@ -1,6 +1,7 @@
 package sifive.fpgashells.ip.microsemi.polarfiredll
 
 import Chisel._
+import chisel3.BlackBox
 import chisel3.experimental.{Analog,attach}
 import freechips.rocketchip.util.{ElaborationArtefacts}
 import freechips.rocketchip.util.GenericParameterizedBundle
diff --git a/src/main/scala/ip/microsemi/polarfire_glitchless_mux/PolarFireGlitchlessMux.scala b/src/main/scala/ip/microsemi/polarfire_glitchless_mux/PolarFireGlitchlessMux.scala
index 993a287..958214c 100644
--- a/src/main/scala/ip/microsemi/polarfire_glitchless_mux/PolarFireGlitchlessMux.scala
+++ b/src/main/scala/ip/microsemi/polarfire_glitchless_mux/PolarFireGlitchlessMux.scala
@@ -1,6 +1,7 @@
 package sifive.fpgashells.ip.microsemi.polarfireglitchlessmux
 
 import Chisel._
+import chisel3.BlackBox
 import chisel3.experimental.{Analog,attach}
 import freechips.rocketchip.util.{ElaborationArtefacts}
 import freechips.rocketchip.util.GenericParameterizedBundle
diff --git a/src/main/scala/ip/microsemi/polarfire_init_monitor/PolarFireInitMonitor.scala b/src/main/scala/ip/microsemi/polarfire_init_monitor/PolarFireInitMonitor.scala
index 3df844d..c5fccbf 100644
--- a/src/main/scala/ip/microsemi/polarfire_init_monitor/PolarFireInitMonitor.scala
+++ b/src/main/scala/ip/microsemi/polarfire_init_monitor/PolarFireInitMonitor.scala
@@ -1,6 +1,7 @@
 package sifive.fpgashells.ip.microsemi.polarfireinitmonitor
 
 import Chisel._
+import chisel3.BlackBox
 import freechips.rocketchip.util.{ElaborationArtefacts}
 import org.chipsalliance.cde.config._
 
diff --git a/src/main/scala/ip/microsemi/polarfire_oscillator/PolarFireOscillator.scala b/src/main/scala/ip/microsemi/polarfire_oscillator/PolarFireOscillator.scala
index 1017d2a..8e1c440 100644
--- a/src/main/scala/ip/microsemi/polarfire_oscillator/PolarFireOscillator.scala
+++ b/src/main/scala/ip/microsemi/polarfire_oscillator/PolarFireOscillator.scala
@@ -1,6 +1,7 @@
 package sifive.fpgashells.ip.microsemi.polarfire_oscillator
 
 import Chisel._
+import chisel3.BlackBox
 import chisel3.experimental.{Analog,attach}
 import freechips.rocketchip.util.{ElaborationArtefacts}
 import freechips.rocketchip.util.GenericParameterizedBundle
diff --git a/src/main/scala/ip/microsemi/polarfire_pcie_rootport/PolarFirePCIeRootPort.scala b/src/main/scala/ip/microsemi/polarfire_pcie_rootport/PolarFirePCIeRootPort.scala
index 919d538..5625bff 100644
--- a/src/main/scala/ip/microsemi/polarfire_pcie_rootport/PolarFirePCIeRootPort.scala
+++ b/src/main/scala/ip/microsemi/polarfire_pcie_rootport/PolarFirePCIeRootPort.scala
@@ -1,6 +1,7 @@
 package sifive.fpgashells.ip.microsemi.polarfirepcierootport
 
 import Chisel._
+import chisel3.BlackBox
 import org.chipsalliance.cde.config._
 import freechips.rocketchip.diplomacy._
 import freechips.rocketchip.amba.axi4._
diff --git a/src/main/scala/ip/microsemi/polarfire_reset/PolarFireReset.scala b/src/main/scala/ip/microsemi/polarfire_reset/PolarFireReset.scala
index a7e2d4f..0945a3e 100644
--- a/src/main/scala/ip/microsemi/polarfire_reset/PolarFireReset.scala
+++ b/src/main/scala/ip/microsemi/polarfire_reset/PolarFireReset.scala
@@ -1,6 +1,7 @@
 package sifive.fpgashells.ip.microsemi.polarfirereset
 
 import Chisel._
+import chisel3.BlackBox
 import chisel3.experimental.{Analog,attach}
 import freechips.rocketchip.util.{ElaborationArtefacts}
 import freechips.rocketchip.util.GenericParameterizedBundle
diff --git a/src/main/scala/ip/microsemi/polarfire_tx_pll/PolarFireTxPLL.scala b/src/main/scala/ip/microsemi/polarfire_tx_pll/PolarFireTxPLL.scala
index fdf9919..02eb5b5 100644
--- a/src/main/scala/ip/microsemi/polarfire_tx_pll/PolarFireTxPLL.scala
+++ b/src/main/scala/ip/microsemi/polarfire_tx_pll/PolarFireTxPLL.scala
@@ -1,6 +1,7 @@
 package sifive.fpgashells.ip.microsemi.polarfiretxpll
 
 import Chisel._
+import chisel3.BlackBox
 import chisel3.experimental.{Analog,attach}
 import freechips.rocketchip.util.{ElaborationArtefacts}
 import freechips.rocketchip.util.GenericParameterizedBundle
diff --git a/src/main/scala/ip/microsemi/polarfire_xcvr_refclk/PolarFireTransceiverRefClk.scala b/src/main/scala/ip/microsemi/polarfire_xcvr_refclk/PolarFireTransceiverRefClk.scala
index 4bee3c3..ef7fdde 100644
--- a/src/main/scala/ip/microsemi/polarfire_xcvr_refclk/PolarFireTransceiverRefClk.scala
+++ b/src/main/scala/ip/microsemi/polarfire_xcvr_refclk/PolarFireTransceiverRefClk.scala
@@ -1,6 +1,7 @@
 package sifive.fpgashells.ip.microsemi.polarfirexcvrrefclk
 
 import Chisel._
+import chisel3.BlackBox
 import chisel3.experimental.{Analog,attach}
 import freechips.rocketchip.util.{ElaborationArtefacts}
 import freechips.rocketchip.util.GenericParameterizedBundle
diff --git a/src/main/scala/ip/xilinx/Unisim.scala b/src/main/scala/ip/xilinx/Unisim.scala
index a350eea..bc38b22 100644
--- a/src/main/scala/ip/xilinx/Unisim.scala
+++ b/src/main/scala/ip/xilinx/Unisim.scala
@@ -1,7 +1,7 @@
 
 package sifive.fpgashells.ip.xilinx
 import Chisel._
-import chisel3.{Input, Output}
+import chisel3.{BlackBox, Input, Output}
 import chisel3.experimental.{Analog, attach, StringParam, RawParam, IntParam, DoubleParam}
 
 import sifive.blocks.devices.pinctrl.{BasePin}
diff --git a/src/main/scala/ip/xilinx/Xilinx.scala b/src/main/scala/ip/xilinx/Xilinx.scala
index 21c69ef..15a5b23 100644
--- a/src/main/scala/ip/xilinx/Xilinx.scala
+++ b/src/main/scala/ip/xilinx/Xilinx.scala
@@ -1,7 +1,7 @@
 package sifive.fpgashells.ip.xilinx
 
 import Chisel._
-import chisel3.{Input, Output}
+import chisel3.{BlackBox, Input, Output}
 import chisel3.experimental.{Analog, attach}
 import chisel3.util.HasBlackBoxInline
 import freechips.rocketchip.util.ElaborationArtefacts
diff --git a/src/main/scala/ip/xilinx/arty100tmig/arty100tmig.scala b/src/main/scala/ip/xilinx/arty100tmig/arty100tmig.scala
index a4061a1..6cb8cee 100644
--- a/src/main/scala/ip/xilinx/arty100tmig/arty100tmig.scala
+++ b/src/main/scala/ip/xilinx/arty100tmig/arty100tmig.scala
@@ -1,6 +1,7 @@
 package sifive.fpgashells.ip.xilinx.arty100tmig
 
 import Chisel._
+import chisel3.BlackBox
 import chisel3.experimental.{Analog,attach}
 import freechips.rocketchip.util.{ElaborationArtefacts}
 import freechips.rocketchip.util.GenericParameterizedBundle
diff --git a/src/main/scala/ip/xilinx/ibufds_gte2/ibufds_gte2.scala b/src/main/scala/ip/xilinx/ibufds_gte2/ibufds_gte2.scala
index 56b9f88..efbd869 100644
--- a/src/main/scala/ip/xilinx/ibufds_gte2/ibufds_gte2.scala
+++ b/src/main/scala/ip/xilinx/ibufds_gte2/ibufds_gte2.scala
@@ -1,6 +1,7 @@
 package sifive.fpgashells.ip.xilinx.ibufds_gte2
 
 import Chisel._
+import chisel3.BlackBox
 
 //IP : xilinx unisim IBUFDS_GTE2
 //Differential Signaling Input Buffer
diff --git a/src/main/scala/ip/xilinx/vc707axi_to_pcie_x1/vc707axi_to_pcie_x1.scala b/src/main/scala/ip/xilinx/vc707axi_to_pcie_x1/vc707axi_to_pcie_x1.scala
index 58f1035..72e8caf 100644
--- a/src/main/scala/ip/xilinx/vc707axi_to_pcie_x1/vc707axi_to_pcie_x1.scala
+++ b/src/main/scala/ip/xilinx/vc707axi_to_pcie_x1/vc707axi_to_pcie_x1.scala
@@ -1,6 +1,7 @@
 package sifive.fpgashells.ip.xilinx.vc707axi_to_pcie_x1
 
 import Chisel._
+import chisel3.BlackBox
 import org.chipsalliance.cde.config._
 import freechips.rocketchip.diplomacy._
 import freechips.rocketchip.amba.axi4._
diff --git a/src/main/scala/ip/xilinx/vc707mig/vc707mig.scala b/src/main/scala/ip/xilinx/vc707mig/vc707mig.scala
index 5091077..6e55b4d 100644
--- a/src/main/scala/ip/xilinx/vc707mig/vc707mig.scala
+++ b/src/main/scala/ip/xilinx/vc707mig/vc707mig.scala
@@ -1,6 +1,7 @@
 package sifive.fpgashells.ip.xilinx.vc707mig
 
 import Chisel._
+import chisel3.BlackBox
 import chisel3.experimental.{Analog,attach}
 import freechips.rocketchip.util.{ElaborationArtefacts}
 import freechips.rocketchip.util.GenericParameterizedBundle
diff --git a/src/main/scala/ip/xilinx/vcu118mig/vcu118mig.scala b/src/main/scala/ip/xilinx/vcu118mig/vcu118mig.scala
index 2842e3f..996b4fb 100644
--- a/src/main/scala/ip/xilinx/vcu118mig/vcu118mig.scala
+++ b/src/main/scala/ip/xilinx/vcu118mig/vcu118mig.scala
@@ -1,6 +1,7 @@
 package sifive.fpgashells.ip.xilinx.vcu118mig
 
 import Chisel._
+import chisel3.BlackBox
 import chisel3.experimental.{Analog,attach}
 import freechips.rocketchip.util.{ElaborationArtefacts}
 import freechips.rocketchip.util.GenericParameterizedBundle
diff --git a/src/main/scala/shell/IOShell.scala b/src/main/scala/shell/IOShell.scala
index 8cdb549..99ae971 100644
--- a/src/main/scala/shell/IOShell.scala
+++ b/src/main/scala/shell/IOShell.scala
@@ -1,7 +1,7 @@
 package sifive.fpgashells.shell
 
 import chisel3._
-import chisel3.experimental.{DataMirror, IO}
+import chisel3.experimental.DataMirror
 import org.chipsalliance.cde.config._
 import freechips.rocketchip.util._
 import freechips.rocketchip.diplomacy._
diff --git a/src/main/scala/shell/Shell.scala b/src/main/scala/shell/Shell.scala
index 3bd8a82..c521766 100644
--- a/src/main/scala/shell/Shell.scala
+++ b/src/main/scala/shell/Shell.scala
@@ -5,10 +5,6 @@ import org.chipsalliance.cde.config._
 import freechips.rocketchip.diplomacy._
 
 import chisel3.experimental.ChiselAnnotation
-import firrtl._
-import firrtl.analyses._
-import firrtl.annotations._
-//import firrtl.ir._
 import freechips.rocketchip.util.DontTouch
 
 case object DesignKey extends Field[Parameters => LazyModule]
diff --git a/src/main/scala/shell/microsemi/VeraShell.scala b/src/main/scala/shell/microsemi/VeraShell.scala
index e7744d6..458642d 100644
--- a/src/main/scala/shell/microsemi/VeraShell.scala
+++ b/src/main/scala/shell/microsemi/VeraShell.scala
@@ -1,7 +1,6 @@
 package sifive.fpgashells.shell.microsemi
 
 import chisel3._
-import chisel3.experimental.IO
 import org.chipsalliance.cde.config._
 import freechips.rocketchip.util._
 import freechips.rocketchip.diplomacy._
diff --git a/src/main/scala/shell/xilinx/ArtyShell.scala b/src/main/scala/shell/xilinx/ArtyShell.scala
index 3128acd..9b88f95 100644
--- a/src/main/scala/shell/xilinx/ArtyShell.scala
+++ b/src/main/scala/shell/xilinx/ArtyShell.scala
@@ -181,6 +181,7 @@ abstract class ArtyShell(implicit val p: Parameters) extends RawModule {
   // Debug JTAG
   //---------------------------------------------------------------------
 
+  /* see chipsalliance/rocket-chip#3279
   def connectDebugJTAG(dut: HasPeripheryDebugModuleImp): SystemJTAGIO = {
 
     require(dut.debug.isDefined, "Connecting JTAG requires that debug module exists")
@@ -229,6 +230,7 @@ abstract class ArtyShell(implicit val p: Parameters) extends RawModule {
 
     djtag
   }
+  */
 
   //---------------------------------------------------------------------
   // UART
diff --git a/src/main/scala/shell/xilinx/PeripheralsVCU118Shell.scala b/src/main/scala/shell/xilinx/PeripheralsVCU118Shell.scala
index e802c7f..ef97e92 100644
--- a/src/main/scala/shell/xilinx/PeripheralsVCU118Shell.scala
+++ b/src/main/scala/shell/xilinx/PeripheralsVCU118Shell.scala
@@ -1,7 +1,7 @@
 package sifive.fpgashells.shell.xilinx
 
 import chisel3._
-import chisel3.experimental.{attach, Analog, IO}
+import chisel3.experimental.{attach, Analog}
 import org.chipsalliance.cde.config._
 import freechips.rocketchip.diplomacy._
 import freechips.rocketchip.tilelink._
diff --git a/src/main/scala/shell/xilinx/SPIFlashXilinxOverlay.scala b/src/main/scala/shell/xilinx/SPIFlashXilinxOverlay.scala
index 01751a5..e9d66d7 100644
--- a/src/main/scala/shell/xilinx/SPIFlashXilinxOverlay.scala
+++ b/src/main/scala/shell/xilinx/SPIFlashXilinxOverlay.scala
@@ -1,7 +1,6 @@
 package sifive.fpgashells.shell.xilinx
 
 import chisel3._
-import chisel3.util.Cat
 import freechips.rocketchip.diplomacy._
 import sifive.fpgashells.shell._
 import sifive.fpgashells.ip.xilinx._
diff --git a/src/main/scala/shell/xilinx/VC707NewShell.scala b/src/main/scala/shell/xilinx/VC707NewShell.scala
index 3ba4f05..a780583 100644
--- a/src/main/scala/shell/xilinx/VC707NewShell.scala
+++ b/src/main/scala/shell/xilinx/VC707NewShell.scala
@@ -1,7 +1,6 @@
 package sifive.fpgashells.shell.xilinx
 
 import chisel3._
-import chisel3.experimental.IO
 import org.chipsalliance.cde.config._
 import freechips.rocketchip.diplomacy._
 import freechips.rocketchip.tilelink._
diff --git a/src/main/scala/shell/xilinx/VC707Shell.scala b/src/main/scala/shell/xilinx/VC707Shell.scala
index 95a86e4..effdea8 100644
--- a/src/main/scala/shell/xilinx/VC707Shell.scala
+++ b/src/main/scala/shell/xilinx/VC707Shell.scala
@@ -63,6 +63,7 @@ trait HasDebugJTAG { this: VC707Shell =>
   val jtag_TDI             = IO(Input(Bool()))
   val jtag_TDO             = IO(Output(Bool()))
 
+  /* See chipsalliance/rocket-chip#3279
   def connectDebugJTAG(dut: HasPeripheryDebugModuleImp, fmcxm105: Boolean = true): SystemJTAGIO = {
   
     require(dut.debug.isDefined, "Connecting JTAG requires that debug module exists")
@@ -123,6 +124,7 @@ trait HasDebugJTAG { this: VC707Shell =>
     dut_ndreset    := dut.debug.get.ndreset
     djtag
   }
+  */
 }
 
 trait HasVC707ChipLink { this: VC707Shell =>
diff --git a/src/main/scala/shell/xilinx/VCU118NewShell.scala b/src/main/scala/shell/xilinx/VCU118NewShell.scala
index c393cbf..7894c8c 100644
--- a/src/main/scala/shell/xilinx/VCU118NewShell.scala
+++ b/src/main/scala/shell/xilinx/VCU118NewShell.scala
@@ -1,7 +1,7 @@
 package sifive.fpgashells.shell.xilinx
 
 import chisel3._
-import chisel3.experimental.{attach, Analog, IO}
+import chisel3.experimental.{attach, Analog}
 import org.chipsalliance.cde.config._
 import freechips.rocketchip.diplomacy._
 import freechips.rocketchip.tilelink._
diff --git a/src/main/scala/shell/xilinx/VCU118Shell.scala b/src/main/scala/shell/xilinx/VCU118Shell.scala
index 005d434..dc482ee 100644
--- a/src/main/scala/shell/xilinx/VCU118Shell.scala
+++ b/src/main/scala/shell/xilinx/VCU118Shell.scala
@@ -204,6 +204,7 @@ abstract class VCU118Shell(implicit val p: Parameters) extends RawModule {
   // Debug JTAG
   //---------------------------------------------------------------------
 
+  /* See chipsalliance/rocket-chip#3279
   def connectDebugJTAG(dut: HasPeripheryDebugModuleImp): SystemJTAGIO = {
     require(dut.debug.isDefined, "Connecting JTAG requires that debug module exists")
     val djtag     = dut.debug.get.systemjtag.get
@@ -220,6 +221,7 @@ abstract class VCU118Shell(implicit val p: Parameters) extends RawModule {
     dut_ndreset    := dut.debug.get.ndreset
     djtag
   }
+  */
 
   //-----------------------------------------------------------------------
   // UART
