diff --git a/src/main/scala/ip/xilinx/bscan/bscan.scala b/src/main/scala/ip/xilinx/bscan/bscan.scala
index 051bce1..519275e 100644
--- a/src/main/scala/ip/xilinx/bscan/bscan.scala
+++ b/src/main/scala/ip/xilinx/bscan/bscan.scala
@@ -6,7 +6,7 @@ import chisel3.experimental.{ExtModule, Analog, attach}
 object JTAGTUNNEL {
   def apply (DUT_TCK: Bool, DUT_TMS: Bool, DUT_TDI: Bool, DUT_TDO:Bool, DUT_TDO_en: Bool): Unit = {
     val inst_jtag_tunnel = Module(new JTAGTUNNEL())
-    DUT_TCK := inst_jtag_tunnel.jtag_tck.asBool()
+    DUT_TCK := inst_jtag_tunnel.jtag_tck.asBool
     DUT_TMS := inst_jtag_tunnel.jtag_tms
     DUT_TDI := inst_jtag_tunnel.jtag_tdi
     inst_jtag_tunnel.jtag_tdo := DUT_TDO
@@ -34,7 +34,7 @@ class BSCANE2 extends ExtModule(Map("JTAG_CHAIN" -> 4)) {
   val UPDATE = IO(Output(Bool()))
 }
 
-class JTAGTUNNEL extends MultiIOModule {
+class JTAGTUNNEL extends Module {
   val jtag_tck: Clock = IO(Output(Clock()))
   val jtag_tms: Bool = IO(Output(Bool()))
   val jtag_tdi: Bool = IO(Output(Bool()))
diff --git a/src/main/scala/shell/xilinx/JTAGDebugBScanOverlay.scala b/src/main/scala/shell/xilinx/JTAGDebugBScanOverlay.scala
index 45eaa1f..89dd50f 100644
--- a/src/main/scala/shell/xilinx/JTAGDebugBScanOverlay.scala
+++ b/src/main/scala/shell/xilinx/JTAGDebugBScanOverlay.scala
@@ -19,7 +19,7 @@ abstract class JTAGDebugBScanXilinxPlacedOverlay(name: String, di: JTAGDebugBSca
 
     JTAGTUNNEL(tmp_tck, tmp_tms, tmp_tdi, tmp_tdo, tmp_tdo_en)
 
-    jtagDebugSink.bundle.TCK := tmp_tck.asClock()
+    jtagDebugSink.bundle.TCK := tmp_tck.asClock
     jtagDebugSink.bundle.TMS := tmp_tms
     jtagDebugSink.bundle.TDI := tmp_tdi
     tmp_tdo := jtagDebugSink.bundle.TDO.data
